scipy/stats/_survival.py

```diff
@@ -1,38 +1,213 @@
+
+        method : str, optional
+            Method used to compute the confidence interval. Options are
+            "Greenwood" and "Exponential Greenwood".
+
+        Returns
+        -------
+        ci : ``ConfidenceInterval``
+            An object with attributes ``low`` and ``high``: arrays of the
+            lower and upper bounds of the confidence interval at unique values
+            of the sample.
+
+        Notes
+        -----
+        Confidence intervals are computed according to the Greenwood formula or
+        the more recent "Exponential Greenwood" formula as described in [1]_.
+
+        References
+        ----------
+        .. [1] Sawyer, Stanley. "The Greenwood and Exponential Greenwood
+               Confidence Intervals in Survival Analysis."
+               https://www.math.wustl.edu/~sawyer/handouts/greenwood.pdf
+
+        """
+        methods = {'greenwood': self._greenwood_ci,
+                   'exponential greenwood': self._exp_greenwood_ci}
+
+        message = f"`method` must be one of {set(methods)}."
+        if method.lower() not in methods:
+            raise ValueError(message)
+
+        message = "`confidence_level` must be a scalar between 0 and 1."
+        confidence_level = np.asarray(confidence_level)[()]
+        if confidence_level.shape or not (0 <= confidence_level <= 1):
+            raise ValueError(message)
+
+        method_fun = methods[method.lower()]
+        low, high = method_fun(confidence_level)
+        return ConfidenceInterval(np.clip(low, 0, 1), np.clip(high, 0, 1))
+
+    def _greenwood_ci(self, confidence_level):
+        sf, d, n = self._sf, self._d, self._n
+        # When n == d, Greenwood's formula divides by zero.
+        # When s != 0, this can be ignored: var == inf, and CI is [0, 1]
+        # When s == 0, this results in NaNs. Produce an informative warning.
+        with np.errstate(divide='ignore', invalid='ignore'):
+            var = sf ** 2 * np.cumsum(d / (n * (n - d)))
+
+        message = ("The variance estimate is undefined at some observations. "
+                   "This is feature of the mathematical formula.")
+        if np.any(np.isnan(var)):
+            warnings.warn(message, RuntimeWarning, stacklevel=2)
+
+        se = np.sqrt(var)
+        z = special.ndtri(1 / 2 + confidence_level / 2)
+
+        low = self.points - z * se
+        high = self.points + z * se
+        return low, high
+
+    def _exp_greenwood_ci(self, confidence_level):
+        sf, d, n = self._sf, self._d, self._n
+
+        var = 1 / np.log(sf) ** 2 * np.cumsum(d / (n * (n - d)))
+
+        se = np.sqrt(var)
+        z = special.ndtri(1 / 2 + confidence_level / 2)
+
+        lnl_points = np.log(-np.log(sf))
+        low = np.exp(-np.exp(lnl_points + z * se))
+        high = np.exp(-np.exp(lnl_points - z * se))
+        if self._kind == "cdf":
+            low, high = 1-high, 1-low
+        return low, high
+
+    def __call__(self, x):
```

Comment:
As we discussed offline, an alternative would be to move this to `ECDFResult` instead. It could also take a name like `ECDFResult.evaluate`. This would return a subset of `ECDFResult` with just `sf` and `cdf` (possibly `x`). Both `sf` and `cdf` would still be `EmpiricalDistributionFunction` without the `__call__`.

Another option here if you wanted to keep the current API on `main` would be to add a `confidence_interval` method to `ECDFResult`. This would return two `ConfidenceInterval` (or another structure, the point being that it would hold both CI for CDF and SF.)

Reply:
Resolving because we decided to remove it for now.